{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import scanpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "#sc.logging.print_versions()\n",
    "#sc.logging.print_memory_usage()\n",
    "#sc.settings.verbosity = 2\n",
    "import os,sys\n",
    "import datetime\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import matplotlib.ticker as mticker\n",
    "from adjustText import adjust_text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib import cm\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.offsetbox import AnchoredOffsetbox, TextArea, HPacker, VPacker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python version: 3.8.12\n"
     ]
    }
   ],
   "source": [
    "# add the utility function folder to PATH\n",
    "sys.path.append(os.path.abspath(\"utility_functions_190403_12h24/\"))\n",
    "\n",
    "from rz_import_statements import *\n",
    "import rz_functions as rz\n",
    "import rz_fig_params # this adjust mpl.rcParams, almost nothing to import, import after scanpy to overwrite rc.Params\n",
    "import rz_utility_spring as srz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('backups_JZ_2022/clean_kidney_304_15tr_25c_batch_corrected_50236x2000_220315_12h14.h5ad') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# overwrite obs with the most recent version\n",
    "filename = 'backups_JZ_2022/corrected_obs_info_50236x32_221012_12h04.npz'\n",
    "encoding = 'latin1'\n",
    "\n",
    "with np.load(filename,encoding=encoding, allow_pickle = True) as f:\n",
    "    obs = pd.DataFrame(**f)\n",
    "adata.obs = obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>library</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>pct_counts_mito</th>\n",
       "      <th>library2</th>\n",
       "      <th>sample</th>\n",
       "      <th>patient</th>\n",
       "      <th>pT stage</th>\n",
       "      <th>seq_date</th>\n",
       "      <th>beads</th>\n",
       "      <th>operation</th>\n",
       "      <th>...</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>no_dblt_no_rbc</th>\n",
       "      <th>sp_cl_43</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>broad_cell_type</th>\n",
       "      <th>color</th>\n",
       "      <th>cell_group</th>\n",
       "      <th>stage_color</th>\n",
       "      <th>patient_color</th>\n",
       "      <th>group_color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N14</td>\n",
       "      <td>449</td>\n",
       "      <td>0.668151</td>\n",
       "      <td>T2_1</td>\n",
       "      <td>T2</td>\n",
       "      <td>P2</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>20_11_12</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>449</td>\n",
       "      <td>True</td>\n",
       "      <td>40</td>\n",
       "      <td>TAM 4</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#9a5ce0</td>\n",
       "      <td>Myeloid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#e87f7f</td>\n",
       "      <td>#85619c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>N14</td>\n",
       "      <td>449</td>\n",
       "      <td>4.23163</td>\n",
       "      <td>T2_1</td>\n",
       "      <td>T2</td>\n",
       "      <td>P2</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>20_11_12</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>449</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>TAM 3</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#943886</td>\n",
       "      <td>Myeloid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#e87f7f</td>\n",
       "      <td>#85619c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>N14</td>\n",
       "      <td>1229</td>\n",
       "      <td>10.6591</td>\n",
       "      <td>T2_1</td>\n",
       "      <td>T2</td>\n",
       "      <td>P2</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>20_11_12</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>1229</td>\n",
       "      <td>True</td>\n",
       "      <td>22</td>\n",
       "      <td>Tumor cells 2</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>#b06c6c</td>\n",
       "      <td>Tumor cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#e87f7f</td>\n",
       "      <td>#db2f2c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>433</th>\n",
       "      <td>N14</td>\n",
       "      <td>432</td>\n",
       "      <td>6.94444</td>\n",
       "      <td>T2_1</td>\n",
       "      <td>T2</td>\n",
       "      <td>P2</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>20_11_12</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>432</td>\n",
       "      <td>True</td>\n",
       "      <td>27</td>\n",
       "      <td>Tumor cells 3</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>#db2f2c</td>\n",
       "      <td>Tumor cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#e87f7f</td>\n",
       "      <td>#db2f2c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>N14</td>\n",
       "      <td>502</td>\n",
       "      <td>4.98008</td>\n",
       "      <td>T2_1</td>\n",
       "      <td>T2</td>\n",
       "      <td>P2</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>20_11_12</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>502</td>\n",
       "      <td>True</td>\n",
       "      <td>28</td>\n",
       "      <td>Tumor vasculature 4</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>#a35927</td>\n",
       "      <td>Endothelium</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#e87f7f</td>\n",
       "      <td>#e07c57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4865536</th>\n",
       "      <td>Tumor0228</td>\n",
       "      <td>683</td>\n",
       "      <td>9.37042</td>\n",
       "      <td>T4_old</td>\n",
       "      <td>T4</td>\n",
       "      <td>P4</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>old</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>683</td>\n",
       "      <td>True</td>\n",
       "      <td>30</td>\n",
       "      <td>TAM 2</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#532b6b</td>\n",
       "      <td>Myeloid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#70c3d4</td>\n",
       "      <td>#85619c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4865584</th>\n",
       "      <td>Tumor0228</td>\n",
       "      <td>1473</td>\n",
       "      <td>6.51731</td>\n",
       "      <td>T4_old</td>\n",
       "      <td>T4</td>\n",
       "      <td>P4</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>old</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>1473</td>\n",
       "      <td>True</td>\n",
       "      <td>30</td>\n",
       "      <td>TAM 2</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#532b6b</td>\n",
       "      <td>Myeloid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#70c3d4</td>\n",
       "      <td>#85619c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4865642</th>\n",
       "      <td>Tumor0228</td>\n",
       "      <td>498</td>\n",
       "      <td>9.43775</td>\n",
       "      <td>T4_old</td>\n",
       "      <td>T4</td>\n",
       "      <td>P4</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>old</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>498</td>\n",
       "      <td>True</td>\n",
       "      <td>37</td>\n",
       "      <td>TAM 3</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#943886</td>\n",
       "      <td>Myeloid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#70c3d4</td>\n",
       "      <td>#85619c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4865726</th>\n",
       "      <td>Tumor0228</td>\n",
       "      <td>421</td>\n",
       "      <td>19.2399</td>\n",
       "      <td>T4_old</td>\n",
       "      <td>T4</td>\n",
       "      <td>P4</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>old</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>421</td>\n",
       "      <td>True</td>\n",
       "      <td>27</td>\n",
       "      <td>Tumor cells 3</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>#db2f2c</td>\n",
       "      <td>Tumor cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#70c3d4</td>\n",
       "      <td>#db2f2c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4866011</th>\n",
       "      <td>Tumor0228</td>\n",
       "      <td>763</td>\n",
       "      <td>3.80079</td>\n",
       "      <td>T4_old</td>\n",
       "      <td>T4</td>\n",
       "      <td>P4</td>\n",
       "      <td>pT3a</td>\n",
       "      <td>old</td>\n",
       "      <td>old</td>\n",
       "      <td>Open</td>\n",
       "      <td>...</td>\n",
       "      <td>763</td>\n",
       "      <td>True</td>\n",
       "      <td>42</td>\n",
       "      <td>Resting/memory T cells</td>\n",
       "      <td>Immune</td>\n",
       "      <td>#4f9e58</td>\n",
       "      <td>Lymphoid cells</td>\n",
       "      <td>#8c1a1a</td>\n",
       "      <td>#70c3d4</td>\n",
       "      <td>#4f9e58</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>50236 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           library total_counts pct_counts_mito library2 sample patient  \\\n",
       "2              N14          449        0.668151     T2_1     T2      P2   \n",
       "19             N14          449         4.23163     T2_1     T2      P2   \n",
       "363            N14         1229         10.6591     T2_1     T2      P2   \n",
       "433            N14          432         6.94444     T2_1     T2      P2   \n",
       "444            N14          502         4.98008     T2_1     T2      P2   \n",
       "...            ...          ...             ...      ...    ...     ...   \n",
       "4865536  Tumor0228          683         9.37042   T4_old     T4      P4   \n",
       "4865584  Tumor0228         1473         6.51731   T4_old     T4      P4   \n",
       "4865642  Tumor0228          498         9.43775   T4_old     T4      P4   \n",
       "4865726  Tumor0228          421         19.2399   T4_old     T4      P4   \n",
       "4866011  Tumor0228          763         3.80079   T4_old     T4      P4   \n",
       "\n",
       "        pT stage  seq_date beads operation  ... n_counts no_dblt_no_rbc  \\\n",
       "2           pT3a  20_11_12   old      Open  ...      449           True   \n",
       "19          pT3a  20_11_12   old      Open  ...      449           True   \n",
       "363         pT3a  20_11_12   old      Open  ...     1229           True   \n",
       "433         pT3a  20_11_12   old      Open  ...      432           True   \n",
       "444         pT3a  20_11_12   old      Open  ...      502           True   \n",
       "...          ...       ...   ...       ...  ...      ...            ...   \n",
       "4865536     pT3a       old   old      Open  ...      683           True   \n",
       "4865584     pT3a       old   old      Open  ...     1473           True   \n",
       "4865642     pT3a       old   old      Open  ...      498           True   \n",
       "4865726     pT3a       old   old      Open  ...      421           True   \n",
       "4866011     pT3a       old   old      Open  ...      763           True   \n",
       "\n",
       "        sp_cl_43               cell_type broad_cell_type    color  \\\n",
       "2             40                   TAM 4          Immune  #9a5ce0   \n",
       "19            37                   TAM 3          Immune  #943886   \n",
       "363           22           Tumor cells 2           Tumor  #b06c6c   \n",
       "433           27           Tumor cells 3           Tumor  #db2f2c   \n",
       "444           28     Tumor vasculature 4     Endothelial  #a35927   \n",
       "...          ...                     ...             ...      ...   \n",
       "4865536       30                   TAM 2          Immune  #532b6b   \n",
       "4865584       30                   TAM 2          Immune  #532b6b   \n",
       "4865642       37                   TAM 3          Immune  #943886   \n",
       "4865726       27           Tumor cells 3           Tumor  #db2f2c   \n",
       "4866011       42  Resting/memory T cells          Immune  #4f9e58   \n",
       "\n",
       "             cell_group stage_color patient_color group_color  \n",
       "2         Myeloid cells     #8c1a1a       #e87f7f     #85619c  \n",
       "19        Myeloid cells     #8c1a1a       #e87f7f     #85619c  \n",
       "363         Tumor cells     #8c1a1a       #e87f7f     #db2f2c  \n",
       "433         Tumor cells     #8c1a1a       #e87f7f     #db2f2c  \n",
       "444         Endothelium     #8c1a1a       #e87f7f     #e07c57  \n",
       "...                 ...         ...           ...         ...  \n",
       "4865536   Myeloid cells     #8c1a1a       #70c3d4     #85619c  \n",
       "4865584   Myeloid cells     #8c1a1a       #70c3d4     #85619c  \n",
       "4865642   Myeloid cells     #8c1a1a       #70c3d4     #85619c  \n",
       "4865726     Tumor cells     #8c1a1a       #70c3d4     #db2f2c  \n",
       "4866011  Lymphoid cells     #8c1a1a       #70c3d4     #4f9e58  \n",
       "\n",
       "[50236 rows x 32 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scale (normalize) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adata.raw.to_adata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 449.]\n",
      " [ 449.]\n",
      " [1229.]\n",
      " [ 432.]\n",
      " [ 502.]]\n",
      "\n",
      "[[10000.001]\n",
      " [10000.   ]\n",
      " [10000.001]\n",
      " [10000.   ]\n",
      " [10000.   ]]\n"
     ]
    }
   ],
   "source": [
    "# turn into counts per 10k\n",
    "print(adata.X[:5,:].sum(axis=1))\n",
    "print()\n",
    "sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)\n",
    "print(adata.X[:5,:].sum(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['TAM 4', 'TAM 3', 'Tumor cells 2', 'Tumor cells 3',\n",
       "       'Tumor vasculature 4', 'TAM 1', 'CD8 T cells', 'TAM 2',\n",
       "       'Resting/memory T cells', 'Cytotoxic T cells',\n",
       "       'Tumor vasculature 2', 'Tumor vasculature 1',\n",
       "       'Non-classical monocytes', 'Mesangial/vSMCs',\n",
       "       'IGHG-high plasma cells', 'Mast cells', 'Tumor vasculature 3',\n",
       "       'Plasma cells', 'Tumor AVR-like vasculature', 'Cycling', 'vSMCs',\n",
       "       'NK cells', 'Classical monocytes', 'B cells', 'Myofibroblasts',\n",
       "       'Regulatory T cells', 'Tumor cells 1', 'Proximal tubule', 'AVR',\n",
       "       'Epithelial progenitor-like cells', 'tAL of LOH',\n",
       "       'Glomerular endothelium', 'Principal cells', 'TAL of LOH', 'DVR',\n",
       "       'Type A-ICs', 'OM Type A-ICs', 'Type B-IC', 'Podocytes', 'DCT/CNT'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select populations of interest\n",
    "adata.obs['cell_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "interest = [ 'Tumor vasculature 4', \n",
    "       'Tumor vasculature 2', 'Tumor vasculature 1', 'Tumor vasculature 3',\n",
    "        'Tumor AVR-like vasculature',  'AVR',\n",
    "       'Glomerular endothelium', 'DVR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmask = adata.obs['cell_type'].isin(interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6394"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(cmask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each cluster, find genes that are statistically significantly higher or lower in cluster x compared to\n",
    "# all other cells collectively\n",
    "\n",
    "#label-free filter to remove very low abundance genes:\n",
    "min_counts = 15\n",
    "min_cells = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6758 genes passing abundance filter\n"
     ]
    }
   ],
   "source": [
    "gmask = srz.filter_abund_genes(adata.X[cmask], min_counts, min_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = 'outputs_JZ_2022/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "mpl.rc('font',family='Arial')\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Volcano plot: tumor vasculature vs healthy vasculature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6394, 33538)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.X[cmask].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "labcol = 'cell_type'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4638 1756\n",
      "1000\n",
      "2000\n",
      "3000\n",
      "4000\n",
      "5000\n",
      "6000\n"
     ]
    }
   ],
   "source": [
    "#perform Mann-Whitney U and get FDR\n",
    "#gmask is the same as before\n",
    "\n",
    "# run \"help(rz.mwu) for details on function parameters\"\n",
    "cmask2 = adata.obs[labcol].isin(['AVR', 'DVR', 'Glomerular endothelium']).values #healthy vasculature\n",
    "cmask1 = adata.obs[labcol].isin(['Tumor vasculature 1', 'Tumor vasculature 2', \n",
    "                                      'Tumor vasculature 4','Tumor vasculature 3',\n",
    "                                     'Tumor AVR-like vasculature']).values #tumor vasculature\n",
    "#cmask = np.repeat(True,adata.shape[0])\n",
    "print(cmask1.sum(), cmask2.sum())\n",
    "\n",
    "# differential gene expression analysis\n",
    "dge = rz.mwu(adata.X[cmask1,:][:,gmask].toarray(),\n",
    "             adata.X[cmask2,:][:,gmask].toarray(),\n",
    "             adata.var_names[gmask])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get fold-change\n",
    "\n",
    "# decide on pseudovalue to add to avoid division by zero, 1 seems to be the scanpy convention\n",
    "pseudo = 1\n",
    "dge['fc'] = (dge['mean1']+pseudo)/(dge['mean2']+pseudo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "dge['log2FC'] = np.log2(dge['fc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>U_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>fdr</th>\n",
       "      <th>mean1</th>\n",
       "      <th>mean2</th>\n",
       "      <th>fc</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>is_DGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "      <td>4095494.5</td>\n",
       "      <td>2.025230e-01</td>\n",
       "      <td>3.054342e-01</td>\n",
       "      <td>0.207651</td>\n",
       "      <td>0.200825</td>\n",
       "      <td>1.005685</td>\n",
       "      <td>0.008179</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HES4</th>\n",
       "      <td>4486332.0</td>\n",
       "      <td>5.791571e-30</td>\n",
       "      <td>1.863783e-28</td>\n",
       "      <td>1.257467</td>\n",
       "      <td>0.401048</td>\n",
       "      <td>1.611270</td>\n",
       "      <td>0.688199</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISG15</th>\n",
       "      <td>4428010.0</td>\n",
       "      <td>9.917145e-11</td>\n",
       "      <td>1.045555e-09</td>\n",
       "      <td>4.647995</td>\n",
       "      <td>4.336897</td>\n",
       "      <td>1.058292</td>\n",
       "      <td>0.081737</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGRN</th>\n",
       "      <td>4448295.0</td>\n",
       "      <td>2.198662e-31</td>\n",
       "      <td>7.392317e-30</td>\n",
       "      <td>0.893393</td>\n",
       "      <td>0.151532</td>\n",
       "      <td>1.644238</td>\n",
       "      <td>0.717419</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNFRSF4</th>\n",
       "      <td>4845289.0</td>\n",
       "      <td>1.145770e-65</td>\n",
       "      <td>1.046367e-63</td>\n",
       "      <td>2.930599</td>\n",
       "      <td>0.511635</td>\n",
       "      <td>2.600230</td>\n",
       "      <td>1.378639</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4L</th>\n",
       "      <td>3799056.0</td>\n",
       "      <td>5.729144e-10</td>\n",
       "      <td>5.507476e-09</td>\n",
       "      <td>1.653327</td>\n",
       "      <td>2.973209</td>\n",
       "      <td>0.667805</td>\n",
       "      <td>-0.582502</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND4</th>\n",
       "      <td>2694553.5</td>\n",
       "      <td>3.334261e-97</td>\n",
       "      <td>5.121122e-95</td>\n",
       "      <td>37.169853</td>\n",
       "      <td>59.805523</td>\n",
       "      <td>0.627737</td>\n",
       "      <td>-0.671769</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND5</th>\n",
       "      <td>3315981.5</td>\n",
       "      <td>7.629344e-32</td>\n",
       "      <td>2.603995e-30</td>\n",
       "      <td>11.927350</td>\n",
       "      <td>17.814468</td>\n",
       "      <td>0.687096</td>\n",
       "      <td>-0.541416</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-ND6</th>\n",
       "      <td>3901898.0</td>\n",
       "      <td>2.706902e-06</td>\n",
       "      <td>1.594877e-05</td>\n",
       "      <td>1.072271</td>\n",
       "      <td>1.772967</td>\n",
       "      <td>0.747312</td>\n",
       "      <td>-0.420217</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MT-CYB</th>\n",
       "      <td>2979944.5</td>\n",
       "      <td>7.882718e-62</td>\n",
       "      <td>6.576717e-60</td>\n",
       "      <td>35.448959</td>\n",
       "      <td>55.816051</td>\n",
       "      <td>0.641526</td>\n",
       "      <td>-0.640421</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6758 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         U_statistic             p           fdr      mean1      mean2  \\\n",
       "NOC2L      4095494.5  2.025230e-01  3.054342e-01   0.207651   0.200825   \n",
       "HES4       4486332.0  5.791571e-30  1.863783e-28   1.257467   0.401048   \n",
       "ISG15      4428010.0  9.917145e-11  1.045555e-09   4.647995   4.336897   \n",
       "AGRN       4448295.0  2.198662e-31  7.392317e-30   0.893393   0.151532   \n",
       "TNFRSF4    4845289.0  1.145770e-65  1.046367e-63   2.930599   0.511635   \n",
       "...              ...           ...           ...        ...        ...   \n",
       "MT-ND4L    3799056.0  5.729144e-10  5.507476e-09   1.653327   2.973209   \n",
       "MT-ND4     2694553.5  3.334261e-97  5.121122e-95  37.169853  59.805523   \n",
       "MT-ND5     3315981.5  7.629344e-32  2.603995e-30  11.927350  17.814468   \n",
       "MT-ND6     3901898.0  2.706902e-06  1.594877e-05   1.072271   1.772967   \n",
       "MT-CYB     2979944.5  7.882718e-62  6.576717e-60  35.448959  55.816051   \n",
       "\n",
       "               fc    log2FC  is_DGE  \n",
       "NOC2L    1.005685  0.008179   False  \n",
       "HES4     1.611270  0.688199   False  \n",
       "ISG15    1.058292  0.081737   False  \n",
       "AGRN     1.644238  0.717419   False  \n",
       "TNFRSF4  2.600230  1.378639    True  \n",
       "...           ...       ...     ...  \n",
       "MT-ND4L  0.667805 -0.582502   False  \n",
       "MT-ND4   0.627737 -0.671769   False  \n",
       "MT-ND5   0.687096 -0.541416   False  \n",
       "MT-ND6   0.747312 -0.420217   False  \n",
       "MT-CYB   0.641526 -0.640421   False  \n",
       "\n",
       "[6758 rows x 8 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Considering genes to be dirrefentially expressed if their fold-change is above 2 and \n",
    "# adjusted p-value is below 0.05\n",
    "\n",
    "fc_ts = 2 # \n",
    "fdr_ts = 0.05 \n",
    "dge['is_DGE'] = (dge['log2FC'].abs()>np.log2(fc_ts))&(dge['fdr']<fdr_ts)\n",
    "dge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#some of the adjusted p-values (fdr) are zero, that causes problems downstream\n",
    "#here I replace such occurences with smallest observed fdr value\n",
    "\n",
    "dge.loc[dge.fdr == 0, 'fdr'] =  sorted(set(dge['fdr']))[1]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.8624920647782902e-282"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(set(dge['fdr']))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating a mask for genes considered as differentialy expressed \n",
    "mask = dge['is_DGE'] == True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>U_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>fdr</th>\n",
       "      <th>mean1</th>\n",
       "      <th>mean2</th>\n",
       "      <th>fc</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>is_DGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TNFRSF4</th>\n",
       "      <td>4845289.0</td>\n",
       "      <td>1.145770e-65</td>\n",
       "      <td>1.046367e-63</td>\n",
       "      <td>2.930599</td>\n",
       "      <td>0.511635</td>\n",
       "      <td>2.600230</td>\n",
       "      <td>1.378639</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VWA1</th>\n",
       "      <td>5267842.5</td>\n",
       "      <td>3.023715e-103</td>\n",
       "      <td>4.983967e-101</td>\n",
       "      <td>6.132065</td>\n",
       "      <td>1.519589</td>\n",
       "      <td>2.830646</td>\n",
       "      <td>1.501131</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HSPG2</th>\n",
       "      <td>5461821.5</td>\n",
       "      <td>4.077799e-121</td>\n",
       "      <td>8.350839e-119</td>\n",
       "      <td>8.092089</td>\n",
       "      <td>2.255594</td>\n",
       "      <td>2.792758</td>\n",
       "      <td>1.481691</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDZK1IP1</th>\n",
       "      <td>3384649.0</td>\n",
       "      <td>1.744535e-117</td>\n",
       "      <td>3.467520e-115</td>\n",
       "      <td>0.367021</td>\n",
       "      <td>3.549050</td>\n",
       "      <td>0.300507</td>\n",
       "      <td>-1.734529</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JUN</th>\n",
       "      <td>3247626.5</td>\n",
       "      <td>5.977698e-47</td>\n",
       "      <td>3.366440e-45</td>\n",
       "      <td>4.181340</td>\n",
       "      <td>10.394905</td>\n",
       "      <td>0.454707</td>\n",
       "      <td>-1.136992</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EVI5L</th>\n",
       "      <td>3735598.0</td>\n",
       "      <td>5.036943e-56</td>\n",
       "      <td>3.583122e-54</td>\n",
       "      <td>0.191240</td>\n",
       "      <td>3.675243</td>\n",
       "      <td>0.254798</td>\n",
       "      <td>-1.972577</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLVAP</th>\n",
       "      <td>6057337.0</td>\n",
       "      <td>2.118203e-209</td>\n",
       "      <td>1.192901e-206</td>\n",
       "      <td>21.928204</td>\n",
       "      <td>6.724150</td>\n",
       "      <td>2.968379</td>\n",
       "      <td>1.569675</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R14A</th>\n",
       "      <td>3383013.5</td>\n",
       "      <td>5.523876e-91</td>\n",
       "      <td>7.466071e-89</td>\n",
       "      <td>0.438580</td>\n",
       "      <td>2.746226</td>\n",
       "      <td>0.384008</td>\n",
       "      <td>-1.380792</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EMP3</th>\n",
       "      <td>3439310.0</td>\n",
       "      <td>1.507806e-60</td>\n",
       "      <td>1.213066e-58</td>\n",
       "      <td>0.873097</td>\n",
       "      <td>3.565980</td>\n",
       "      <td>0.410229</td>\n",
       "      <td>-1.285499</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIOX</th>\n",
       "      <td>3609635.0</td>\n",
       "      <td>5.305872e-105</td>\n",
       "      <td>9.194124e-103</td>\n",
       "      <td>0.104017</td>\n",
       "      <td>1.872587</td>\n",
       "      <td>0.384329</td>\n",
       "      <td>-1.379588</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          U_statistic              p            fdr      mean1      mean2  \\\n",
       "TNFRSF4     4845289.0   1.145770e-65   1.046367e-63   2.930599   0.511635   \n",
       "VWA1        5267842.5  3.023715e-103  4.983967e-101   6.132065   1.519589   \n",
       "HSPG2       5461821.5  4.077799e-121  8.350839e-119   8.092089   2.255594   \n",
       "PDZK1IP1    3384649.0  1.744535e-117  3.467520e-115   0.367021   3.549050   \n",
       "JUN         3247626.5   5.977698e-47   3.366440e-45   4.181340  10.394905   \n",
       "...               ...            ...            ...        ...        ...   \n",
       "EVI5L       3735598.0   5.036943e-56   3.583122e-54   0.191240   3.675243   \n",
       "PLVAP       6057337.0  2.118203e-209  1.192901e-206  21.928204   6.724150   \n",
       "PPP1R14A    3383013.5   5.523876e-91   7.466071e-89   0.438580   2.746226   \n",
       "EMP3        3439310.0   1.507806e-60   1.213066e-58   0.873097   3.565980   \n",
       "MIOX        3609635.0  5.305872e-105  9.194124e-103   0.104017   1.872587   \n",
       "\n",
       "                fc    log2FC  is_DGE  \n",
       "TNFRSF4   2.600230  1.378639    True  \n",
       "VWA1      2.830646  1.501131    True  \n",
       "HSPG2     2.792758  1.481691    True  \n",
       "PDZK1IP1  0.300507 -1.734529    True  \n",
       "JUN       0.454707 -1.136992    True  \n",
       "...            ...       ...     ...  \n",
       "EVI5L     0.254798 -1.972577    True  \n",
       "PLVAP     2.968379  1.569675    True  \n",
       "PPP1R14A  0.384008 -1.380792    True  \n",
       "EMP3      0.410229 -1.285499    True  \n",
       "MIOX      0.384329 -1.379588    True  \n",
       "\n",
       "[100 rows x 8 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge = dge[mask]\n",
    "dge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#saving the file\n",
    "dge.to_excel(outdir + 'tumor_vs_healthy_endo.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot volcano plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I am plotting a volcano plot where significant and differentially expressed genes are highlighted in blue or red (down- or up-regulated) with their number in the dataset, to avoid clutter plotting all the gene names. I am also saving a dictionary to decipher which number is which gene, and later on I can edit the image in Illustrator selecting the genes I wish to emphasize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = np.log2(dge['fc'])\n",
    "y = -np.log10(dge['fdr'])\n",
    "\n",
    "s = 3 #dotsize\n",
    "c = '0.2'#dotcolor\n",
    "alpha = 0.5 #dot transparency\n",
    "lw = 0\n",
    "# select thresholds for what to consider \"hits\"\n",
    "fc_ts = 2 # \n",
    "fdr_ts = 0.05 # convention \n",
    "\n",
    "xlabel = 'log2[(Tumor vasculature)\\n/(Healthy vasculature)]'\n",
    "ylabel = '-log10(adj. p-value)'\n",
    "\n",
    "# plot dots\n",
    "a,fig,gs = rz.startfig(14,10)\n",
    "a.scatter(x,y,c=c,s=s,lw=lw,alpha=alpha)\n",
    "\n",
    "# center fold-change around zero:\n",
    "xmax = abs(np.array(a.get_xlim())).max()\n",
    "a.set_xlim(-xmax,xmax)\n",
    "\n",
    "\n",
    "# plot thresholds:\n",
    "(left,right) = a.get_xlim()\n",
    "top = a.get_ylim()[1]\n",
    "logfcts = np.log2(fc_ts)\n",
    "logfdrts = -np.log10(fdr_ts)\n",
    "\n",
    "\n",
    "# save hits as a list, color hits in red:\n",
    "hitmask = dge['is_DGE']\n",
    "hits = dge.index[hitmask]\n",
    "\n",
    "# also split hit in those up left and up right\n",
    "upleft = hits[dge.loc[hits,'fc']<1]\n",
    "upright = hits[dge.loc[hits,'fc']>1]\n",
    "\n",
    "numberTogeneUp = {} # Save genes up in text file made from dict to make your life easier in illustator/keynote\n",
    "numberTogeneDo = {} # Save genes do in text file made from dict to make your life easier in illustator/keynote\n",
    "\n",
    "textsUp = []\n",
    "textsDo = []\n",
    "\n",
    "iup = 1\n",
    "ido = 1\n",
    "for gene,m,n in zip(hits,x[hitmask],y[hitmask]):\n",
    "    if m > 0: \n",
    "        a.scatter(m,n,lw=lw,s=s*4.5,c='r',alpha=1, zorder = 6)\n",
    "        numberTogeneUp[gene] = iup\n",
    "        textsUp.append(plt.text(m,n,iup,fontsize=12.5, zorder = 8))\n",
    "        iup += 1\n",
    "        \n",
    "    elif m < 0: \n",
    "        a.scatter(m,n,lw=lw,s=s*4.5,c='#008ad4',alpha=1, zorder = 6)\n",
    "        numberTogeneDo[gene] = ido\n",
    "        textsDo.append(plt.text(m,n,ido,fontsize=12.5, zorder = 8))\n",
    "        ido += 1\n",
    "\n",
    "# Annotate significant DGE with numbers in plot (essential that adjust_text() is called last)\n",
    "adjust_text(textsUp, \n",
    "            arrowprops=dict(arrowstyle='-', color='k', zorder = 4, lw = 0.75),\n",
    "            precision = 0.0000000001,\n",
    "            force_text=(0.2, 0.45), \n",
    "            )\n",
    "adjust_text(textsDo, \n",
    "            arrowprops=dict(arrowstyle='-', color='k', zorder = 4, lw = 0.75), \n",
    "            precision = 0.0000000001,\n",
    "            force_text=(0.2, 0.45), \n",
    "            )\n",
    "\n",
    "a.set_xlabel(xlabel, fontsize=16)\n",
    "a.set_ylabel(ylabel, fontsize =16)\n",
    "\n",
    "a.set_ylim(0,290)\n",
    "\n",
    "\n",
    "rz.showspines(a,left=True,bottom=True)\n",
    "\n",
    "\n",
    "\n",
    "#plt.savefig(outdir+'volcano_tumor_vasculature_vs_healthy.pdf', dpi=600, bbox_inches='tight')\n",
    "\n",
    "# print the number of hits:\n",
    "print('# gene up left',len(upleft))\n",
    "print('# genes up right',len(upright))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>U_statistic</th>\n",
       "      <th>p</th>\n",
       "      <th>fdr</th>\n",
       "      <th>mean1</th>\n",
       "      <th>mean2</th>\n",
       "      <th>fc</th>\n",
       "      <th>log2FC</th>\n",
       "      <th>is_DGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VWA1</th>\n",
       "      <td>5267842.5</td>\n",
       "      <td>3.023715e-103</td>\n",
       "      <td>4.983967e-101</td>\n",
       "      <td>6.132065</td>\n",
       "      <td>1.519589</td>\n",
       "      <td>2.830646</td>\n",
       "      <td>1.501131</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INHBB</th>\n",
       "      <td>4840706.0</td>\n",
       "      <td>6.350843e-81</td>\n",
       "      <td>7.035901e-79</td>\n",
       "      <td>2.280245</td>\n",
       "      <td>0.062248</td>\n",
       "      <td>3.088021</td>\n",
       "      <td>1.626682</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP7</th>\n",
       "      <td>6868030.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.862492e-282</td>\n",
       "      <td>97.637604</td>\n",
       "      <td>26.680046</td>\n",
       "      <td>3.563491</td>\n",
       "      <td>1.833291</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPARCL1</th>\n",
       "      <td>6372203.5</td>\n",
       "      <td>2.541425e-285</td>\n",
       "      <td>2.862492e-282</td>\n",
       "      <td>25.330284</td>\n",
       "      <td>4.908820</td>\n",
       "      <td>4.456099</td>\n",
       "      <td>2.155781</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPRY1</th>\n",
       "      <td>6014161.0</td>\n",
       "      <td>1.519694e-204</td>\n",
       "      <td>7.335778e-202</td>\n",
       "      <td>22.339014</td>\n",
       "      <td>5.879164</td>\n",
       "      <td>3.392711</td>\n",
       "      <td>1.762439</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LAMA4</th>\n",
       "      <td>4992530.0</td>\n",
       "      <td>3.095876e-93</td>\n",
       "      <td>4.269781e-91</td>\n",
       "      <td>2.577341</td>\n",
       "      <td>0.201170</td>\n",
       "      <td>2.978213</td>\n",
       "      <td>1.574447</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP3</th>\n",
       "      <td>5294444.0</td>\n",
       "      <td>5.425069e-137</td>\n",
       "      <td>1.410101e-134</td>\n",
       "      <td>10.878271</td>\n",
       "      <td>0.306484</td>\n",
       "      <td>9.091784</td>\n",
       "      <td>3.184563</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANGPT2</th>\n",
       "      <td>5313972.5</td>\n",
       "      <td>1.338440e-138</td>\n",
       "      <td>3.618071e-136</td>\n",
       "      <td>5.204752</td>\n",
       "      <td>0.257791</td>\n",
       "      <td>4.933055</td>\n",
       "      <td>2.302481</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STC1</th>\n",
       "      <td>4964428.0</td>\n",
       "      <td>5.117912e-95</td>\n",
       "      <td>7.358904e-93</td>\n",
       "      <td>3.460555</td>\n",
       "      <td>0.230238</td>\n",
       "      <td>3.625765</td>\n",
       "      <td>1.858285</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENPP2</th>\n",
       "      <td>4775476.0</td>\n",
       "      <td>7.040058e-57</td>\n",
       "      <td>5.061352e-55</td>\n",
       "      <td>5.856410</td>\n",
       "      <td>0.833809</td>\n",
       "      <td>3.738891</td>\n",
       "      <td>1.902610</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KCNE3</th>\n",
       "      <td>4857973.0</td>\n",
       "      <td>1.279367e-83</td>\n",
       "      <td>1.490683e-81</td>\n",
       "      <td>2.543161</td>\n",
       "      <td>0.081056</td>\n",
       "      <td>3.277500</td>\n",
       "      <td>1.712596</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRSS23</th>\n",
       "      <td>5204659.0</td>\n",
       "      <td>5.896004e-105</td>\n",
       "      <td>9.961300e-103</td>\n",
       "      <td>5.166786</td>\n",
       "      <td>0.896823</td>\n",
       "      <td>3.251114</td>\n",
       "      <td>1.700934</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VWF</th>\n",
       "      <td>6328114.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.862492e-282</td>\n",
       "      <td>11.613110</td>\n",
       "      <td>0.447000</td>\n",
       "      <td>8.716732</td>\n",
       "      <td>3.123787</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A1</th>\n",
       "      <td>5499445.0</td>\n",
       "      <td>2.258353e-136</td>\n",
       "      <td>5.652575e-134</td>\n",
       "      <td>8.429471</td>\n",
       "      <td>1.458041</td>\n",
       "      <td>3.836173</td>\n",
       "      <td>1.939668</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A2</th>\n",
       "      <td>5434349.0</td>\n",
       "      <td>4.635228e-132</td>\n",
       "      <td>1.044162e-129</td>\n",
       "      <td>6.394942</td>\n",
       "      <td>1.149595</td>\n",
       "      <td>3.440155</td>\n",
       "      <td>1.782474</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INSR</th>\n",
       "      <td>5798400.5</td>\n",
       "      <td>1.582110e-205</td>\n",
       "      <td>8.224537e-203</td>\n",
       "      <td>7.661953</td>\n",
       "      <td>0.533378</td>\n",
       "      <td>5.648936</td>\n",
       "      <td>2.497979</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLVAP</th>\n",
       "      <td>6057337.0</td>\n",
       "      <td>2.118203e-209</td>\n",
       "      <td>1.192901e-206</td>\n",
       "      <td>21.928204</td>\n",
       "      <td>6.724150</td>\n",
       "      <td>2.968379</td>\n",
       "      <td>1.569675</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         U_statistic              p            fdr      mean1      mean2  \\\n",
       "VWA1       5267842.5  3.023715e-103  4.983967e-101   6.132065   1.519589   \n",
       "INHBB      4840706.0   6.350843e-81   7.035901e-79   2.280245   0.062248   \n",
       "IGFBP7     6868030.0   0.000000e+00  2.862492e-282  97.637604  26.680046   \n",
       "SPARCL1    6372203.5  2.541425e-285  2.862492e-282  25.330284   4.908820   \n",
       "SPRY1      6014161.0  1.519694e-204  7.335778e-202  22.339014   5.879164   \n",
       "LAMA4      4992530.0   3.095876e-93   4.269781e-91   2.577341   0.201170   \n",
       "IGFBP3     5294444.0  5.425069e-137  1.410101e-134  10.878271   0.306484   \n",
       "ANGPT2     5313972.5  1.338440e-138  3.618071e-136   5.204752   0.257791   \n",
       "STC1       4964428.0   5.117912e-95   7.358904e-93   3.460555   0.230238   \n",
       "ENPP2      4775476.0   7.040058e-57   5.061352e-55   5.856410   0.833809   \n",
       "KCNE3      4857973.0   1.279367e-83   1.490683e-81   2.543161   0.081056   \n",
       "PRSS23     5204659.0  5.896004e-105  9.961300e-103   5.166786   0.896823   \n",
       "VWF        6328114.0   0.000000e+00  2.862492e-282  11.613110   0.447000   \n",
       "COL4A1     5499445.0  2.258353e-136  5.652575e-134   8.429471   1.458041   \n",
       "COL4A2     5434349.0  4.635228e-132  1.044162e-129   6.394942   1.149595   \n",
       "INSR       5798400.5  1.582110e-205  8.224537e-203   7.661953   0.533378   \n",
       "PLVAP      6057337.0  2.118203e-209  1.192901e-206  21.928204   6.724150   \n",
       "\n",
       "               fc    log2FC  is_DGE  \n",
       "VWA1     2.830646  1.501131    True  \n",
       "INHBB    3.088021  1.626682    True  \n",
       "IGFBP7   3.563491  1.833291    True  \n",
       "SPARCL1  4.456099  2.155781    True  \n",
       "SPRY1    3.392711  1.762439    True  \n",
       "LAMA4    2.978213  1.574447    True  \n",
       "IGFBP3   9.091784  3.184563    True  \n",
       "ANGPT2   4.933055  2.302481    True  \n",
       "STC1     3.625765  1.858285    True  \n",
       "ENPP2    3.738891  1.902610    True  \n",
       "KCNE3    3.277500  1.712596    True  \n",
       "PRSS23   3.251114  1.700934    True  \n",
       "VWF      8.716732  3.123787    True  \n",
       "COL4A1   3.836173  1.939668    True  \n",
       "COL4A2   3.440155  1.782474    True  \n",
       "INSR     5.648936  2.497979    True  \n",
       "PLVAP    2.968379  1.569675    True  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dge[dge['log2FC']>1.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TNFRSF4', 'VWA1', 'HSPG2', 'ACKR1', 'RGS5', 'IVNS1ABP', 'CYTOR',\n",
       "       'MIR4435-2HG', 'INHBB', 'CXCR4', 'COL8A1', 'MGLL', 'IGFBP7', 'SPARCL1',\n",
       "       'SPRY1', 'ESM1', 'SPARC', 'LAMA4', 'GJA1', 'IGFBP3', 'GRB10', 'CXorf36',\n",
       "       'ANGPT2', 'STC1', 'CA2', 'ENPP2', 'COL15A1', 'TP53I11', 'KCNE3',\n",
       "       'PRSS23', 'MCAM', 'THY1', 'VIM', 'UNC5B', 'HTRA1', 'VWF', 'NDUFA4L2',\n",
       "       'MLEC', 'RGCC', 'EDNRB', 'COL4A1', 'COL4A2', 'CCL3', 'INSR', 'PLVAP'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upright"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TNFRSF4': 1,\n",
       " 'VWA1': 2,\n",
       " 'HSPG2': 3,\n",
       " 'ACKR1': 4,\n",
       " 'RGS5': 5,\n",
       " 'IVNS1ABP': 6,\n",
       " 'CYTOR': 7,\n",
       " 'MIR4435-2HG': 8,\n",
       " 'INHBB': 9,\n",
       " 'CXCR4': 10,\n",
       " 'COL8A1': 11,\n",
       " 'MGLL': 12,\n",
       " 'IGFBP7': 13,\n",
       " 'SPARCL1': 14,\n",
       " 'SPRY1': 15,\n",
       " 'ESM1': 16,\n",
       " 'SPARC': 17,\n",
       " 'LAMA4': 18,\n",
       " 'GJA1': 19,\n",
       " 'IGFBP3': 20,\n",
       " 'GRB10': 21,\n",
       " 'CXorf36': 22,\n",
       " 'ANGPT2': 23,\n",
       " 'STC1': 24,\n",
       " 'CA2': 25,\n",
       " 'ENPP2': 26,\n",
       " 'COL15A1': 27,\n",
       " 'TP53I11': 28,\n",
       " 'KCNE3': 29,\n",
       " 'PRSS23': 30,\n",
       " 'MCAM': 31,\n",
       " 'THY1': 32,\n",
       " 'VIM': 33,\n",
       " 'UNC5B': 34,\n",
       " 'HTRA1': 35,\n",
       " 'VWF': 36,\n",
       " 'NDUFA4L2': 37,\n",
       " 'MLEC': 38,\n",
       " 'RGCC': 39,\n",
       " 'EDNRB': 40,\n",
       " 'COL4A1': 41,\n",
       " 'COL4A2': 42,\n",
       " 'CCL3': 43,\n",
       " 'INSR': 44,\n",
       " 'PLVAP': 45}"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numberTogeneUp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_up = pd.DataFrame.from_dict([numberTogeneUp])\n",
    "df_do = pd.DataFrame.from_dict([numberTogeneDo])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_up.to_csv(outdir + 'tumor_vasc_up.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_do.to_csv(outdir + 'tumor_vasc_do.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
